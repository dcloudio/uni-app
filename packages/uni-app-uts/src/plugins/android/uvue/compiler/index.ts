import { extend } from '@vue/shared'
import { type CompilerError, baseParse } from '@vue/compiler-core'

import { isAppUVueNativeTag } from '@dcloudio/uni-shared'
import {
  createUniVueTransformAssetUrls,
  getBaseNodeTransforms,
  transformTapToClick,
} from '@dcloudio/uni-cli-shared'
import './runtimeHelpers'

import type { CodegenResult, TemplateCompilerOptions } from './options'
import { generate } from './codegen'
import {
  type DirectiveTransform,
  type NodeTransform,
  transform,
} from './transform'
import { transformIf } from './transforms/vIf'
import { transformFor } from './transforms/vFor'
import { transformModel } from './transforms/vModel'
import { transformShow } from './transforms/vShow'
import { transformVText } from './transforms/vText'
import { transformInterpolation } from './transforms/transformInterpolation'
import { transformText } from './transforms/transformText'
import { transformOn } from './transforms/vOnWithModifier'
import { transformBind } from './transforms/vBind'
import { transformSlotOutlet } from './transforms/transformSlotOutlet'
import { transformObjectExpression } from './transforms/transformObjectExpression'
import { transformExpression } from './transforms/transformExpression'
import { transformElements } from './transforms/transformElements'
import { transformStyle } from './transforms/transformStyle'
import { transformVHtml } from './transforms/vHtml'
import { transformMemo } from './transforms/vMemo'
import { transformOnce } from './transforms/vOnce'
import {
  type RawSourceMap,
  SourceMapConsumer,
  SourceMapGenerator,
} from 'source-map-js'
import { trackSlotScopes, trackVForSlotScopes } from './transforms/vSlot'
import { transformElement } from './transforms/transformElement'
import {
  createAssetUrlTransformWithOptions,
  normalizeOptions,
} from '../sfc/compiler/template/transformAssetUrl'
import { createSrcsetTransformWithOptions } from '../sfc/compiler/template/transformSrcset'

export type TransformPreset = [
  NodeTransform[],
  Record<string, DirectiveTransform>
]

export function getBaseTransformPreset(
  prefixIdentifiers?: boolean
): TransformPreset {
  return [
    [
      transformOnce,
      transformIf,
      transformMemo,
      transformFor,
      // order is important
      trackVForSlotScopes,
      transformVHtml,
      transformExpression,
      transformSlotOutlet,
      transformElement,
      trackSlotScopes,
      transformText,
      transformTapToClick,
      transformInterpolation,
      transformObjectExpression,
      transformElements,
      transformStyle,
    ] as any,
    {
      on: transformOn,
      bind: transformBind,
      model: transformModel,
      show: transformShow,
      text: transformVText,
    } as any,
  ]
}

export function compile(
  template: string,
  options: TemplateCompilerOptions
): CodegenResult {
  options.rootDir = options.rootDir || ''
  options.targetLanguage = options.targetLanguage || 'kotlin'
  options.prefixIdentifiers =
    'prefixIdentifiers' in options
      ? options.prefixIdentifiers
      : options.mode === 'module'

  wrapOptionsLog(template, options)

  const isNativeTag =
    options?.isNativeTag ||
    function (tag: string) {
      return (
        isAppUVueNativeTag(tag) ||
        !!options.parseUTSComponent?.(tag, options.targetLanguage!)
      )
    }
  const ast = baseParse(template, {
    comments: false,
    isNativeTag,
    onError: options.onError,
  })
  const [nodeTransforms, directiveTransforms] = getBaseTransformPreset(
    options.prefixIdentifiers
  )

  // 重要不能传入base
  const assetOptions = normalizeOptions(createUniVueTransformAssetUrls(''))

  transform(
    ast,
    extend({}, options, {
      nodeTransforms: [
        ...nodeTransforms,
        createAssetUrlTransformWithOptions(assetOptions),
        createSrcsetTransformWithOptions(assetOptions),
        ...getBaseNodeTransforms('/'),
        ...(options.nodeTransforms || []), // user transforms
      ],
      directiveTransforms: extend(
        {},
        directiveTransforms,
        options.directiveTransforms || {} // user transforms
      ),
    })
  )

  const result = generate(ast, options)

  // inMap should be the map produced by ./parse.ts which is a simple line-only
  // mapping. If it is present, we need to adjust the final map and errors to
  // reflect the original line numbers.
  if (options.inMap) {
    if (options.sourceMap) {
      result.map = mapLines(options.inMap, result.map!)
    }
    // if (result.errors.length) {
    //   patchErrors(errors, source, inMap)
    // }
  }

  return result
}

function mapLines(oldMap: RawSourceMap, newMap: RawSourceMap): RawSourceMap {
  if (!oldMap) return newMap
  if (!newMap) return oldMap

  const oldMapConsumer = new SourceMapConsumer(oldMap)
  const newMapConsumer = new SourceMapConsumer(newMap)
  const mergedMapGenerator = new SourceMapGenerator()

  newMapConsumer.eachMapping((m) => {
    if (m.originalLine == null) {
      return
    }

    const origPosInOldMap = oldMapConsumer.originalPositionFor({
      line: m.originalLine,
      column: m.originalColumn ?? 0,
    })

    if (origPosInOldMap.source == null) {
      return
    }

    mergedMapGenerator.addMapping({
      generated: {
        line: m.generatedLine,
        column: m.generatedColumn,
      },
      original: {
        line: origPosInOldMap.line, // map line
        // use current column, since the oldMap produced by @vue/compiler-sfc
        // does not
        column: m.originalColumn ?? 0,
      },
      source: origPosInOldMap.source,
      name: origPosInOldMap.name,
    })
  })

  // source-map's type definition is incomplete
  const generator = mergedMapGenerator as any
  ;(oldMapConsumer as any).sources.forEach((sourceFile: string) => {
    generator._sources.add(sourceFile)
    const sourceContent = oldMapConsumer.sourceContentFor(sourceFile)
    if (sourceContent != null) {
      mergedMapGenerator.setSourceContent(sourceFile, sourceContent)
    }
  })

  generator._sourceRoot = oldMap.sourceRoot
  generator._file = oldMap.file
  return generator.toJSON()
}

function wrapOptionsLog(source: string, options: TemplateCompilerOptions) {
  const { onWarn, onError, inMap } = options
  if (inMap && inMap.sourcesContent?.length) {
    if (onWarn || onError) {
      const originalSource = inMap.sourcesContent![0]
      const offset = originalSource.indexOf(source)
      const lineOffset =
        originalSource.slice(0, offset).split(/\r?\n/).length - 1

      if (onWarn) {
        options.onWarn = (err: CompilerError) => {
          patchError(err, lineOffset, offset)
          onWarn(err)
        }
      }
      if (onError) {
        options.onError = (err: CompilerError) => {
          patchError(err, lineOffset, offset)
          onError(err)
        }
      }
    }
  }
}

function patchError(err: CompilerError, lineOffset: number, offset: number) {
  if (err.loc) {
    err.loc.start.line += lineOffset
    err.loc.start.offset += offset
    if (err.loc.end !== err.loc.start) {
      err.loc.end.line += lineOffset
      err.loc.end.offset += offset
    }
  }
}
