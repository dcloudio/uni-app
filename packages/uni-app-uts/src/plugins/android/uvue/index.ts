import path from 'path'
import fs from 'fs-extra'

import type { Plugin } from 'vite'
import type { SFCBlock } from '@vue/compiler-sfc'

import { isString } from '@vue/shared'
import {
  genUTSClassName,
  normalizePath,
  parseVueRequest,
  resolveAppVue,
} from '@dcloudio/uni-cli-shared'

import {
  getDescriptor,
  getResolvedOptions,
  getSrcDescriptor,
} from './descriptorCache'
import { isVue, transformUniCloudMixinDataCom } from '../utils'

import { transformStyle } from './code/style'

import { transformMain } from './sfc/main'

export function uniAppUVuePlugin(): Plugin {
  const options = getResolvedOptions()
  const appVue = resolveAppVue(process.env.UNI_INPUT_DIR)
  function isAppVue(id: string) {
    return normalizePath(id) === appVue
  }
  return {
    name: 'uni:app-uvue',
    apply: 'build',
    async resolveId(id) {
      // serve sub-part requests (*?vue) as virtual modules
      if (parseVueRequest(id).query.vue) {
        return id
      }
    },

    load(id) {
      const { filename, query } = parseVueRequest(id)
      // select corresponding block for sub-part virtual modules
      if (query.vue) {
        if (query.src) {
          return fs.readFileSync(filename, 'utf-8')
        }
        const descriptor = getDescriptor(filename, options)!
        let block: SFCBlock | null | undefined
        if (query.type === 'style') {
          block = descriptor.styles[query.index!]
        } else if (query.index != null) {
          block = descriptor.customBlocks[query.index]
        }
        if (block) {
          return {
            code: block.content,
            map: block.map as any,
          }
        }
      }
    },
    async transform(code, id) {
      const { filename, query } = parseVueRequest(id)
      if (!isVue(filename)) {
        return
      }
      if (!query.vue) {
        // main request
        return transformMain(
          transformUniCloudMixinDataCom(code),
          filename,
          {
            ...options,
            componentType: isAppVue(filename)
              ? 'app'
              : query.type === 'page'
              ? 'page'
              : 'component',
          },
          this
        )
      } else {
        // sub block request
        const descriptor = query.src
          ? getSrcDescriptor(filename)!
          : getDescriptor(filename, options)!

        if (query.type === 'style') {
          return transformStyle(
            code,
            descriptor,
            Number(query.index),
            options,
            this,
            filename
          )
        }
      }
    },
    generateBundle(_, bundle) {
      // 遍历vue文件，填充style，尽量减少全局变量
      Object.keys(bundle).forEach((name) => {
        const file = bundle[name]
        if (
          file &&
          file.type === 'asset' &&
          isVueFile(file.fileName) &&
          isString(file.source)
        ) {
          let fileName = normalizePath(file.fileName)
          if (process.env.UNI_APP_X_TSC === 'true') {
            fileName = fileName.replace('.ts', '')
          }
          const className =
            process.env.UNI_COMPILE_TARGET === 'ext-api'
              ? // components/map/map.vue => UniMap
                genUTSClassName(
                  path.basename(fileName),
                  options.classNamePrefix
                )
              : genUTSClassName(fileName, options.classNamePrefix)

          const classNameComment = `/*${className}Styles*/`
          if (file.source.includes(classNameComment)) {
            const styleAssetName = fileName + '.style.uts'
            const styleAsset = bundle[styleAssetName]
            if (
              styleAsset &&
              styleAsset.type === 'asset' &&
              isString(styleAsset.source)
            ) {
              file.source = file.source.replace(
                classNameComment,
                styleAsset.source.replace('export ', '')
              )
              delete bundle[styleAssetName]
            }
          }
        }
      })
    },
  }
}

function isVueFile(filename: string) {
  return (
    filename.endsWith('.uvue') ||
    filename.endsWith('.vue') ||
    filename.endsWith('.uvue.ts') ||
    filename.endsWith('.vue.ts')
  )
}
