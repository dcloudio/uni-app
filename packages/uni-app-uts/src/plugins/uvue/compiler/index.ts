import { extend } from '@vue/shared'
import {
  baseParse,
  trackSlotScopes,
  trackVForSlotScopes,
  transformElement,
  transformExpression,
} from '@vue/compiler-core'

import { isAppUVueNativeTag } from '@dcloudio/uni-shared'
import { transformTapToClick } from '@dcloudio/uni-cli-shared'
import './runtimeHelpers'

import { CodegenResult, CompilerOptions } from './options'
import { generate } from './codegen'
import { DirectiveTransform, NodeTransform, transform } from './transform'
import { transformIf } from './transforms/vIf'
import { transformFor } from './transforms/vFor'
import { transformModel } from './transforms/vModel'
import { transformShow } from './transforms/vShow'
import { transformVText } from './transforms/vText'
import { transformInterpolation } from './transforms/transformInterpolation'
import { transformText } from './transforms/transformText'
import { transformOn } from './transforms/vOn'
import { transformBind } from './transforms/vBind'
import { transformSlotOutlet } from './transforms/transformSlotOutlet'

export type TransformPreset = [
  NodeTransform[],
  Record<string, DirectiveTransform>
]

export function getBaseTransformPreset(
  prefixIdentifiers?: boolean
): TransformPreset {
  return [
    [
      transformIf,
      transformFor,
      // order is important
      trackVForSlotScopes,
      transformExpression,
      transformSlotOutlet,
      transformElement,
      trackSlotScopes,
      transformText,
      transformTapToClick,
      transformInterpolation,
    ] as any,
    {
      on: transformOn,
      bind: transformBind,
      model: transformModel,
      show: transformShow,
      text: transformVText,
    } as any,
  ]
}

export function compile(
  template: string,
  options: CompilerOptions
): CodegenResult {
  const ast = baseParse(template, {
    isNativeTag(tag) {
      return isAppUVueNativeTag(tag)
    },
  })
  const [nodeTransforms, directiveTransforms] = getBaseTransformPreset(
    options.prefixIdentifiers
  )

  transform(
    ast,
    extend({}, options, {
      prefixIdentifiers: options.prefixIdentifiers,
      nodeTransforms: [
        ...nodeTransforms,
        ...(options.nodeTransforms || []), // user transforms
      ],
      directiveTransforms: extend(
        {},
        directiveTransforms,
        options.directiveTransforms || {} // user transforms
      ),
    })
  )

  return generate(ast, options)
}
