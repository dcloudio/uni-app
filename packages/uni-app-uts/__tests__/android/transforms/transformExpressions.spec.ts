// import { PatchFlagNames, PatchFlags } from '@vue/shared'
import {
  BindingTypes,
  ConstantTypes,
  type DirectiveNode,
  type ElementNode,
  type InterpolationNode,
  NodeTypes,
  baseParse as parse,
} from '@vue/compiler-core'
import { transform } from '../../../src/plugins/android/uvue/compiler/transform'
import { transformIf } from '../../../src/plugins/android/uvue/compiler/transforms/vIf'
import { transformExpression } from '../../../src/plugins/android/uvue/compiler/transforms/transformExpression'

import { compile as baseCompile } from '../../../src/plugins/android/uvue/compiler'
import type { TemplateCompilerOptions } from '../../../src/plugins/android/uvue/compiler/options'

function parseWithExpressionTransform(
  template: string,
  options: TemplateCompilerOptions = {}
) {
  const ast = parse(template)
  transform(ast, {
    prefixIdentifiers: true,
    nodeTransforms: [transformIf, transformExpression],
    ...options,
  })
  return ast.children[0]
}

describe('compiler: expression transform', () => {
  test('interpolation (root)', () => {
    const node = parseWithExpressionTransform(`{{ foo }}`) as InterpolationNode
    expect(node.content).toMatchObject({
      type: NodeTypes.SIMPLE_EXPRESSION,
      content: `_ctx.foo`,
    })
  })

  test('empty interpolation', () => {
    const node = parseWithExpressionTransform(`{{}}`) as InterpolationNode
    const node2 = parseWithExpressionTransform(`{{ }}`) as InterpolationNode
    const node3 = parseWithExpressionTransform(
      `<view>{{ }}</view>`
    ) as ElementNode

    const objectToBeMatched = {
      type: NodeTypes.SIMPLE_EXPRESSION,
      content: ``,
    }
    expect(node.content).toMatchObject(objectToBeMatched)
    expect(node2.content).toMatchObject(objectToBeMatched)
    expect((node3.children[0] as InterpolationNode).content).toMatchObject(
      objectToBeMatched
    )
  })

  test('interpolation (children)', () => {
    const el = parseWithExpressionTransform(
      `<view>{{ foo }}</view>`
    ) as ElementNode
    const node = el.children[0] as InterpolationNode
    expect(node.content).toMatchObject({
      type: NodeTypes.SIMPLE_EXPRESSION,
      content: `_ctx.foo`,
    })
  })

  test('interpolation (complex)', () => {
    const el = parseWithExpressionTransform(
      `<view>{{ foo + bar(baz.qux) }}</view>`
    ) as ElementNode
    const node = el.children[0] as InterpolationNode
    expect(node.content).toMatchObject({
      type: NodeTypes.COMPOUND_EXPRESSION,
      children: [
        { content: `_ctx.foo` },
        ` + `,
        { content: `_ctx.bar` },
        `(`,
        { content: `_ctx.baz` },
        `.`,
        { content: `qux` },
        `)`,
      ],
    })
  })

  test('directive value', () => {
    const node = parseWithExpressionTransform(
      `<view v-foo:arg="baz"/>`
    ) as ElementNode
    const arg = (node.props[0] as DirectiveNode).arg!
    expect(arg).toMatchObject({
      type: NodeTypes.SIMPLE_EXPRESSION,
      content: `arg`,
    })
    const exp = (node.props[0] as DirectiveNode).exp!
    expect(exp).toMatchObject({
      type: NodeTypes.SIMPLE_EXPRESSION,
      content: `_ctx.baz`,
    })
  })

  test('dynamic directive arg', () => {
    const node = parseWithExpressionTransform(
      `<view v-foo:[arg]="baz"/>`
    ) as ElementNode
    const arg = (node.props[0] as DirectiveNode).arg!
    expect(arg).toMatchObject({
      type: NodeTypes.SIMPLE_EXPRESSION,
      content: `_ctx.arg`,
    })
    const exp = (node.props[0] as DirectiveNode).exp!
    expect(exp).toMatchObject({
      type: NodeTypes.SIMPLE_EXPRESSION,
      content: `_ctx.baz`,
    })
  })

  test('should prefix complex expressions', () => {
    const node = parseWithExpressionTransform(
      `{{ foo(baz + 1, { key: kuz }) }}`
    ) as InterpolationNode
    // should parse into compound expression
    expect(node.content).toMatchObject({
      type: NodeTypes.COMPOUND_EXPRESSION,
      children: [
        {
          content: `_ctx.foo`,
          loc: {
            source: `foo`,
            start: {
              offset: 3,
              line: 1,
              column: 4,
            },
            end: {
              offset: 6,
              line: 1,
              column: 7,
            },
          },
        },
        `(`,
        {
          content: `_ctx.baz`,
          loc: {
            source: `baz`,
            start: {
              offset: 7,
              line: 1,
              column: 8,
            },
            end: {
              offset: 10,
              line: 1,
              column: 11,
            },
          },
        },
        ` + 1, { key: `,
        {
          content: `_ctx.kuz`,
          loc: {
            source: `kuz`,
            start: {
              offset: 23,
              line: 1,
              column: 24,
            },
            end: {
              offset: 26,
              line: 1,
              column: 27,
            },
          },
        },
        ` })`,
      ],
    })
  })

  test('should not prefix whitelisted globals', () => {
    const node = parseWithExpressionTransform(
      `{{ Math.max(1, 2) }}`
    ) as InterpolationNode
    expect(node.content).toMatchObject({
      type: NodeTypes.COMPOUND_EXPRESSION,
      children: [{ content: `Math` }, `.`, { content: `max` }, `(1, 2)`],
    })

    const node2 = parseWithExpressionTransform(
      `{{ uni.env.USER_DATA_PATH }}`
    ) as InterpolationNode
    expect(node2.content).toMatchObject({
      type: NodeTypes.COMPOUND_EXPRESSION,
      children: [
        { content: `uni` },
        `.`,
        { content: `env` },
        `.`,
        { content: `USER_DATA_PATH` },
      ],
    })
  })

  test('should not prefix reserved literals', () => {
    function assert(exp: string) {
      const node = parseWithExpressionTransform(
        `{{ ${exp} }}`
      ) as InterpolationNode
      expect(node.content).toMatchObject({
        type: NodeTypes.SIMPLE_EXPRESSION,
        content: exp,
      })
    }
    assert(`true`)
    assert(`false`)
    assert(`null`)
    assert(`this`)
  })

  test('should not prefix id of a function declaration', () => {
    const node = parseWithExpressionTransform(
      `{{ function foo() { return bar } }}`
    ) as InterpolationNode
    expect(node.content).toMatchObject({
      type: NodeTypes.COMPOUND_EXPRESSION,
      children: [
        `function `,
        { content: `foo` },
        `() { return `,
        { content: `_ctx.bar` },
        ` }`,
      ],
    })
  })

  test('should not prefix params of a function expression', () => {
    const node = parseWithExpressionTransform(
      `{{ foo => foo + bar }}`
    ) as InterpolationNode
    expect(node.content).toMatchObject({
      type: NodeTypes.COMPOUND_EXPRESSION,
      children: [
        { content: `foo` },
        ` => `,
        { content: `foo` },
        ` + `,
        { content: `_ctx.bar` },
      ],
    })
  })

  test('should prefix default value of a function expression param', () => {
    const node = parseWithExpressionTransform(
      `{{ (foo = baz) => foo + bar }}`
    ) as InterpolationNode
    expect(node.content).toMatchObject({
      type: NodeTypes.COMPOUND_EXPRESSION,
      children: [
        `(`,
        { content: `foo` },
        ` = `,
        { content: `_ctx.baz` },
        `) => `,
        { content: `foo` },
        ` + `,
        { content: `_ctx.bar` },
      ],
    })
  })

  test('should not prefix function param destructuring', () => {
    const node = parseWithExpressionTransform(
      `{{ ({ foo }) => foo + bar }}`
    ) as InterpolationNode
    expect(node.content).toMatchObject({
      type: NodeTypes.COMPOUND_EXPRESSION,
      children: [
        `({ `,
        { content: `foo` },
        ` }) => `,
        { content: `foo` },
        ` + `,
        { content: `_ctx.bar` },
      ],
    })
  })

  test('function params should not affect out of scope identifiers', () => {
    const node = parseWithExpressionTransform(
      `{{ { a: foo => foo, b: foo } }}`
    ) as InterpolationNode
    expect(node.content).toMatchObject({
      type: NodeTypes.COMPOUND_EXPRESSION,
      children: [
        `{ a: `,
        { content: `foo` },
        ` => `,
        { content: `foo` },
        `, b: `,
        { content: `_ctx.foo` },
        ` }`,
      ],
    })
  })

  test('should prefix default value of function param destructuring', () => {
    const node = parseWithExpressionTransform(
      `{{ ({ foo = bar }) => foo + bar }}`
    ) as InterpolationNode
    expect(node.content).toMatchObject({
      type: NodeTypes.COMPOUND_EXPRESSION,
      children: [
        `({ `,
        { content: `foo` },
        ` = `,
        { content: `_ctx.bar` },
        ` }) => `,
        { content: `foo` },
        ` + `,
        { content: `_ctx.bar` },
      ],
    })
  })
  test('should not prefix an object property key', () => {
    const node = parseWithExpressionTransform(
      `{{ { foo() { baz() }, value: bar } }}`
    ) as InterpolationNode
    expect(node.content).toMatchObject({
      type: NodeTypes.COMPOUND_EXPRESSION,
      children: [
        `{ foo() { `,
        { content: `_ctx.baz` },
        `() }, value: `,
        { content: `_ctx.bar` },
        ` }`,
      ],
    })
  })

  test('should not duplicate object key with same name as value', () => {
    const node = parseWithExpressionTransform(
      `{{ { foo: foo } }}`
    ) as InterpolationNode
    expect(node.content).toMatchObject({
      type: NodeTypes.COMPOUND_EXPRESSION,
      children: [`{ foo: `, { content: `_ctx.foo` }, ` }`],
    })
  })

  test('should prefix a computed object property key', () => {
    const node = parseWithExpressionTransform(
      `{{ { [foo]: bar } }}`
    ) as InterpolationNode
    expect(node.content).toMatchObject({
      type: NodeTypes.COMPOUND_EXPRESSION,
      children: [
        `{ [`,
        { content: `_ctx.foo` },
        `]: `,
        { content: `_ctx.bar` },
        ` }`,
      ],
    })
  })

  test('should prefix object property shorthand value', () => {
    const node = parseWithExpressionTransform(
      `{{ { foo } }}`
    ) as InterpolationNode
    expect(node.content).toMatchObject({
      type: NodeTypes.COMPOUND_EXPRESSION,
      children: [`{ foo: `, { content: `_ctx.foo` }, ` }`],
    })
  })

  test('should not prefix id in a member expression', () => {
    const node = parseWithExpressionTransform(
      `{{ foo.bar.baz }}`
    ) as InterpolationNode
    expect(node.content).toMatchObject({
      type: NodeTypes.COMPOUND_EXPRESSION,
      children: [
        { content: `_ctx.foo` },
        `.`,
        { content: `bar` },
        `.`,
        { content: `baz` },
      ],
    })
  })

  test('should prefix computed id in a member expression', () => {
    const node = parseWithExpressionTransform(
      `{{ foo[bar][baz] }}`
    ) as InterpolationNode
    expect(node.content).toMatchObject({
      type: NodeTypes.COMPOUND_EXPRESSION,
      children: [
        { content: `_ctx.foo` },
        `[`,
        { content: `_ctx.bar` },
        `][`,
        { content: '_ctx.baz' },
        `]`,
      ],
    })
  })

  test('should handle parse error', () => {
    // const onError = vi.fn()
    // parseWithExpressionTransform(`{{ a( }}`, { onError })
    // expect(onError.mock.calls[0][0].message).toMatch(
    //   `Error parsing JavaScript expression: Unexpected token`
    // )
  })

  test('should prefix in assignment', () => {
    const node = parseWithExpressionTransform(
      `{{ x = 1 }}`
    ) as InterpolationNode
    expect(node.content).toMatchObject({
      type: NodeTypes.COMPOUND_EXPRESSION,
      children: [{ content: `_ctx.x` }, ` = 1`],
    })
  })

  test('should prefix in assignment pattern', () => {
    const node = parseWithExpressionTransform(
      `{{ { x, y: [z] } = obj }}`
    ) as InterpolationNode
    expect(node.content).toMatchObject({
      type: NodeTypes.COMPOUND_EXPRESSION,
      children: [
        `{ x: `,
        { content: `_ctx.x` },
        `, y: [`,
        { content: `_ctx.z` },
        `] } = `,
        { content: `_ctx.obj` },
      ],
    })
  })

  // #8295
  test('should treat floating point number literals as constant', () => {
    const node = parseWithExpressionTransform(
      `{{ [1, 2.1] }}`
    ) as InterpolationNode
    expect(node.content).toMatchObject({
      constType: ConstantTypes.CAN_STRINGIFY,
    })
  })

  describe('ES Proposals support', () => {
    test('bigInt', () => {
      const node = parseWithExpressionTransform(
        `{{ 13000n }}`
      ) as InterpolationNode
      expect(node.content).toMatchObject({
        type: NodeTypes.SIMPLE_EXPRESSION,
        content: `13000n`,
        isStatic: false,
        constType: ConstantTypes.CAN_STRINGIFY,
      })
    })

    test('nullish coalescing', () => {
      const node = parseWithExpressionTransform(
        `{{ a ?? b }}`
      ) as InterpolationNode
      expect(node.content).toMatchObject({
        type: NodeTypes.COMPOUND_EXPRESSION,
        children: [{ content: `_ctx.a` }, ` ?? `, { content: `_ctx.b` }],
      })
    })

    test('optional chaining', () => {
      const node = parseWithExpressionTransform(
        `{{ a?.b?.c }}`
      ) as InterpolationNode
      expect(node.content).toMatchObject({
        type: NodeTypes.COMPOUND_EXPRESSION,
        children: [
          { content: `_ctx.a` },
          `?.`,
          { content: `b` },
          `?.`,
          { content: `c` },
        ],
      })
    })

    test('Enabling additional plugins', () => {
      // enabling pipeline operator to replace filters:
      // const node = parseWithExpressionTransform(`{{ a |> uppercase }}`, {
      //   expressionPlugins: [
      //     [
      //       'pipelineOperator',
      //       {
      //         proposal: 'minimal',
      //       },
      //     ],
      //   ],
      // }) as InterpolationNode
      // expect(node.content).toMatchObject({
      //   type: NodeTypes.COMPOUND_EXPRESSION,
      //   children: [
      //     { content: `_ctx.a` },
      //     ` |> `,
      //     { content: `_ctx.uppercase` },
      //   ],
      // })
    })
  })

  describe('bindingMetadata', () => {
    const bindingMetadata = {
      props: BindingTypes.PROPS,
      setup: BindingTypes.SETUP_MAYBE_REF,
      setupConst: BindingTypes.SETUP_CONST,
      data: BindingTypes.DATA,
      options: BindingTypes.OPTIONS,
      reactive: BindingTypes.SETUP_REACTIVE_CONST,
      literal: BindingTypes.LITERAL_CONST,
      isNaN: BindingTypes.SETUP_REF,
    }

    function compileWithBindingMetadata(
      template: string,
      options?: TemplateCompilerOptions
    ) {
      return baseCompile(template, {
        prefixIdentifiers: true,
        bindingMetadata,
        ...options,
      })
    }

    test('non-inline mode', () => {
      const { code } = compileWithBindingMetadata(
        `<view>{{ props }} {{ setup }} {{ data }} {{ options }} {{ isNaN }}</view>`
      )
      expect(code).toMatch(`_ctx.props`)
      expect(code).toMatch(`_ctx.setup`)
      expect(code).toMatch(`_ctx.isNaN`)
      expect(code).toMatch(`_ctx.data`)
      expect(code).toMatch(`_ctx.options`)
      // expect(code).toMatch(`_ctx, _cache, $props, $setup, $data, $options`)
      expect(code).toMatchSnapshot()
    })

    test('should not prefix temp variable of for...in', () => {
      const { code } = compileWithBindingMetadata(
        `<view @click="() => {
          for (const x in list) {
            log(x)
          }         
        }"/>`
      )
      expect(code).not.toMatch(`_ctx.x`)
      expect(code).toMatchSnapshot()
    })

    test('should not prefix temp variable of for...of', () => {
      const { code } = compileWithBindingMetadata(
        `<view @click="() => {
          for (const x of list) {
            log(x)
          }         
        }"/>`
      )
      expect(code).not.toMatch(`_ctx.x`)
      expect(code).toMatchSnapshot()
    })

    test('should not prefix temp variable of for loop', () => {
      const { code } = compileWithBindingMetadata(
        `<view @click="() => {
          for (let i = 0; i < list.length; i++) {
            log(i)
          }         
        }"/>`
      )
      expect(code).not.toMatch(`_ctx.i`)
      expect(code).toMatchSnapshot()
    })

    test('inline mode', () => {
      const { code } = compileWithBindingMetadata(
        `<view>{{ props }} {{ setup }} {{ setupConst }} {{ data }} {{ options }} {{ isNaN }}</view>`,
        { inline: true }
      )
      expect(code).toMatch(`_ctx.props`)
      expect(code).toMatch(`unref(setup)`)
      expect(code).toMatch(`toDisplayString(setupConst)`)
      expect(code).toMatch(`_ctx.data`)
      expect(code).toMatch(`_ctx.options`)
      expect(code).toMatch(`isNaN.value`)
      expect(code).toMatchSnapshot()
    })

    test('literal const handling', () => {
      const { code } = compileWithBindingMetadata(
        `<view>{{ literal }}</view>`,
        {
          inline: true,
        }
      )
      expect(code).toMatch(`toDisplayString(literal)`)
      // #7973 should skip patch for literal const
      // TODO
      // expect(code).not.toMatch(
      //   `${PatchFlags.TEXT} /* ${PatchFlagNames[PatchFlags.TEXT]} */`
      // )
    })

    test('literal const handling， non-inline mode', () => {
      const { code } = compileWithBindingMetadata(`<view>{{ literal }}</view>`)
      expect(code).toMatch(`toDisplayString(_ctx.literal)`)
      // #7973 should skip patch for literal const
      // TODO
      // expect(code).not.toMatch(
      //   `${PatchFlags.TEXT} /* ${PatchFlagNames[PatchFlags.TEXT]} */`
      // )
    })

    test('reactive const handling', () => {
      // const { code } = compileWithBindingMetadata(`<view>{{ reactive }}</view>`, {
      //   inline: true,
      // })
      // #7973 should not skip patch for reactive const
      // TODO
      // expect(code).toMatch(
      //   `${PatchFlags.TEXT} /* ${PatchFlagNames[PatchFlags.TEXT]} */`
      // )
    })
  })
})
