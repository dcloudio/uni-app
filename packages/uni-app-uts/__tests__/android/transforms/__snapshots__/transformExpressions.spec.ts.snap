// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compiler: expression transform bindingMetadata inline mode 1`] = `"_cE("view", null, _tD(_ctx.props) + " " + _tD(unref(setup)) + " " + _tD(setupConst) + " " + _tD(_ctx.data) + " " + _tD(_ctx.options) + " " + _tD(isNaN.value), 1 /* TEXT */)"`;

exports[`compiler: expression transform bindingMetadata non-inline mode 1`] = `"_cE("view", null, _tD(_ctx.props) + " " + _tD(_ctx.setup) + " " + _tD(_ctx.data) + " " + _tD(_ctx.options) + " " + _tD(_ctx.isNaN), 1 /* TEXT */)"`;

exports[`compiler: expression transform bindingMetadata should not prefix temp variable of for loop 1`] = `
"_cE("view", _uM({
  onClick: () => {
          for (let i = 0; i < _ctx.list.length; i++) {
            _ctx.log(i)
          }         
        }
}), null, 8 /* PROPS */, ["onClick"])"
`;

exports[`compiler: expression transform bindingMetadata should not prefix temp variable of for...in 1`] = `
"_cE("view", _uM({
  onClick: () => {
          for (const x in _ctx.list) {
            _ctx.log(x)
          }         
        }
}), null, 8 /* PROPS */, ["onClick"])"
`;

exports[`compiler: expression transform bindingMetadata should not prefix temp variable of for...of 1`] = `
"_cE("view", _uM({
  onClick: () => {
          for (const x of _ctx.list) {
            _ctx.log(x)
          }         
        }
}), null, 8 /* PROPS */, ["onClick"])"
`;
